### DISCUSSION TITLE
Custom command definition changes

### DISCUSSION THREAD

**Discussion by fdncred** [nushell/nushell#6458] [Polls]
A recent PR was landed that changes custom commands from this syntax:
```
# brackets
def foo [x: int] {
    $x + 100
}
```
to this:
```
# parens
def foo (x: int) {
    $x + 100
}
```

**Comment by CBenoit** [comment]
I voted for parens. One argument I saw in favor of square brackets over parenthesizes was the following:

> since you don't use parens to call functions, for example: `cd foo` would call `cd` and pass `foo`

However, we don't use square brackets either when calling functions, and thus parenthesizes are not any worse in the first place IMO. Following this reasoning, I think it’s strictly better than square brackets because it looks familiar to most programmers.

That being said, I want to bring attention to a third alternative favoring consistency and symmetry in nu language. I’m wondering if using _nothing at all_ was considered.

```
def foo 
    a  : int@"complete a", # Parameter a
    b? : string,  # Optional b
{ 
    ...
}

def bar y : int {
   y + 100
}

def "more arguments" a : int, b : int, c : int {
   …
}
```

This may be less familiar than parenthesizes, but since it’s symmetric it should be easier to remember than square brackets. Maybe it’s not great for visual separation but doesn’t look too bad to me.

What do you think?

**Reply by Zoybean** [reply]
I forget how this works presently, but how would you specify the return type with this syntax

**Reply by CBenoit** [reply]
You currently can’t specify the return type in nu. This syntax could support that with `->` but that would be yet another keyword.
I prefer your suggestion below about symmetry between command parameter definition and block/closure definition.

**Comment by Zoybean** [comment]
Tl;dr: I propose that while changes to command definition syntax are being discussed, the options can also be assessed in terms of how feasible a corresponding closure syntax would be, as I think there would be value in these areas of the language looking similar.


One thing I'd like (if it's possible) is symmetry between command parameter definition and block/closure definition syntax (See also #4115). A closure is very conceptually similar to an anonymous or nested command, so I think it would aid readability if they used similar syntax.
Currently we've got `def name [params]{body}` for command definition and `{|params| body}` for closure definition, which don't do a good job of hinting their similarity. Possibly we could have `[params]{body}` for closures to make the syntax align with command definition (I don't think this syntax would be ambiguous, I don't think square brackets are used anywhere else). But if we change command definition to `def name (params){body}`, the corresponding closure syntax would be something like `(params){body}`, which could be ambiguous given the existing meaning of parentheses in the language.

**Reply by Zoybean** [reply]
Whoops, I said I don't think anything else uses `[]` but of course table literals use that syntax. Regardless, I think my main point still stands; it would be nice if the command definition syntax was able to be the same as the closure definition syntax, whatever that is

**Comment by pepdiz** [comment]
In my humble opion:

Parens never , at all !    Parens are used in nushell to group commands and thus identify an executable structure, a param list is never an executable structure.

I agree with Zoybean about [using | | ](https://github.com/nushell/nushell/discussions/6458#discussioncomment-3555557) the same way when defining arguments in code blocks but the problem is code blocks are very special kind in nushell since it seems code blocks only accept one parameter and it must be type Any.

So I think the best approach is keep brackets [ ]  as brackets define lists in nushell and that is a parameter list, a list. Also in terms of reflection and introspection when getting params in brackets you get a list in terms of nushell datatypes but if params get inside parens you get nothing but an error in terms of nushell datatypes.  

So in terms of reflection and internal coherence, brackets are the best option.

**Reply by Zoybean** [reply]
What you said about brackets as a list struck true immediately, but then I realised, isn't it more like a record? A heterogeneous list with named elements. That said, I think I would dislike `{}` for function parameters, especially since the same is used for the function body.

**Reply by pepdiz** [reply]
I agree with you, even when a record can be used for parameters with some advantages like named parameters I think it would be confuse for the reason you said and also becouse a parameter list is ultimately a list nor a record, parameters are positional by nature and in a record position is irrelevant.  As a compromise I would accept a list of records for parameter lists ;-)  but maybe too verbose:

`def f [ {name: a, default: 2, optional: true, flag: true, type: int} {name: b} {name: c , completion: my-completion }] { ... } `

taking advantage of default values for record fields when not present (here if flag is not present its assumed to be false, type is assumed to be Any , same for optional or default}

against:

`def f [ --a: int = 2, b, c@completion} ] { ... } `

Actual mode is based in special syntax while the list of records relays and internal structures and types of nushell what makes it more verbose and maybe clumsy but better for introspection and handling data as code.  

The verbose syntax can be resolved with syntax sugar in the parser.  The parse can translate:

`def f [ a: int = 2, b:Any, c@completion} ] { ... } `

into

`def f [ {name: a, default: 2, optional: true, flag: true, type: int} {name: b} {name: c , completion: my-completion }] { ... } `

and 

```
def greet [
  name: string     # The name of the person to greet
  --age (-a): int   # The age of the person
] {
  echo $name $age
}
```
into 

```
def greet [ {name: name, type: string, desc: "The name of the person to greet"}
           {name: age, shortname: a, type: int, desc: "The age of the person"} 
] {
  echo $name $age
}
```
and 
```
def greet [...name: string] {
  echo "hello all:"
  for $n in $name {
    echo $n
  }
}
```
into
```
def greet [ {name: name, type:string, argList: true} ] {
  echo "hello all:"
  for $n in $name {
    echo $n
  }
}
```
and
```
def greet [name?: string] {
  echo "hello" $name | str collect
}
```
into
```
def greet [ {name: name, optional: true, type: string} ] {
  echo "hello" $name | str collect
}
```
and
```
def my-command [animal: string@animals] { print $animal }
```
into
```
def my-command [ {name: animal, type: string, completion: animals} ] { print $animal }
```
or something similar

The use of syntax sugar gives you extra features for free, now you can decorate parameters and it's easy to extend the decorations (or parameter's properties) for example adding annotations and you do that in a simple straitght way using expresiveness of nushell language and own internal structure.

Also you avoid a cumbersone syntax in definining commands,  currently you use a kind of list to define parameters but it is not a real list but a special syntax similar to list at first sight, in the command definition 

`def greet [  n: string   # The name of the person to greet  --age (-a): int   # The age of the person ] {...}` 

cannot be written because is a invalid syntax, since comment extends to the end of line and parameter list is not closed (nushell waits for expect end of list ']'),  so your are forced to write it in separated lines:

```
def greet [  n: string   # The name of the person to greet  
   --age (-a): int   # The age of the person 
] {...}
```
if commands were a list the listem items should be (assuming comment is parsed out as it really is):

- n: 
- string   
- --age 
- -a: 
- int   

rather than what it's pretended:

- n: string   # The name of the person to greet 
-  --age (-a): int   # The age of the person 

and this is exactly what happend:

```
〉[ n: string # The name of the person to greet
::: --age (-a) : int # The age of the person
::: ]
╭───┬────────╮
│ 0 │ n:     │
│ 1 │ string │
│ 2 │ --age  │
│ 3 │ -a     │
│ 4 │ :      │
│ 5 │ int    │
╰───┴────────╯
```

Note that argument list is a special syntax and the parser does not parse the argument list the same way any other list.

Parsing the argument list as a list fails with an error:

```
〉[ n: string # The name of the person to greet   
::: --age (-a): int # The age of the person
::: ]
Error: nu::parser::unclosed_delimiter (link)

  × Unclosed delimiter.
   ╭─[entry #26:1:1]
 1 │ [ n: string # The name of the person to greet
 2 │ --age (-a): int # The age of the person
   ·            ▲
   ·            ╰── unclosed )
 3 │ ]
   ╰────
```
because the parser doesn't recognize a ): as a parens terminator and thus (-a): is a unclosed parens, so you are forced to separate the colon from the end paren with a space, thus getting an extra item in the list , the colon

while the argument list in a command definition is parsed diferently because is a special syntax, thus no error:
```
〉def greet [  n: string   # The name of the person to greet
:::    --age (-a): int   # The age of the person
::: ] {echo $n}
〉greet "e"
e
```

so you are forced to treat parameter lists as a special syntax and not as a proper list.

Also you may need , as item list separator to improve readability, this is a valid definition `def f [b a:int = 2] { echo $a $b }` of a command f accepting two parameter, b and a, but it's clearer to write it as `def f [b, a:int = 2] { echo $a $b }` which is also a valid definition  (thus introducing more ambiguity to improve readability)

This simply doesn't happend if parameter are records and parameter list is a proper list of records
```
def greet [ {name: n, type: string, desc: "The name of the person to greet"} 
             {name: age, shortname: a, type: int, desc: "The age of the person"} ] {...}
```

allways gets the correct list items:
- {name: n, type: string, desc: "The name of the person to greet"} 
- {name: age, shortname: a, type: int, desc: "The age of the person"}

**Comment by adamijak** [comment]
@pepdiz Why not to use `()` for parameters and `{}` for executable block?

**Reply by pepdiz** [reply]
because parens () are used for [subexpresions](https://www.nushell.sh/book/variables_and_subexpressions.html#subexpressions), and subexpressions are evaluable elements that return a value (or executable commands that return a value, if you prefer) and parameters are not executable or evaluable elements and don't return a value, 'returning a value' is a nonsense  in a parameter definition.
